src/Interpreters/ExpressionActions.h

```diff
@@ -184,102 +102,233 @@ class ActionsDAG
 
     const Node & addInput(std::string name, DataTypePtr type);
     const Node & addInput(ColumnWithTypeAndName column);
     const Node & addColumn(ColumnWithTypeAndName column);
     const Node & addAlias(const std::string & name, std::string alias, bool can_replace = false);
-    const Node & addArrayJoin(const std::string & source_name, std::string result_name, std::string unique_column_name);
+    const Node & addArrayJoin(const std::string & source_name, std::string result_name);
     const Node & addFunction(
             const FunctionOverloadResolverPtr & function,
             const Names & argument_names,
             std::string result_name,
-            bool compile_expressions);
+            const Context & context);
+
+    /// Call addAlias several times.
+    void addAliases(const NamesWithAliases & aliases);
+    /// Adds alias actions and removes unused columns from index.
+    void project(const NamesWithAliases & projection);
 
-    ExpressionActionsPtr buildExpressions(const Context & context);
+    /// Removes column from index.
+    void removeColumn(const std::string & column_name);
+    /// If column is not in index, try to find it in nodes and insert back into index.
+    bool tryRestoreColumn(const std::string & column_name);
+
+    void projectInput() { project_input = true; }
+    void removeUnusedActions(const Names & required_names);
+    ExpressionActionsPtr buildExpressions() const;
+
+    /// Splits actions into two parts. Returned half may be swapped with ARRAY JOIN.
+    /// Returns nullptr if no actions may be moved before ARRAY JOIN.
+    ActionsDAGPtr splitActionsBeforeArrayJoin(const NameSet & array_joined_columns);
+
+    bool hasArrayJoin() const;
+    bool empty() const; /// If actions only contain inputs.
+    bool projectedOutput() const { return projected_output; } /// Remove all columns which are not in inputs from block.
+
+    ActionsDAGPtr clone() const;
 
 private:
     Node & addNode(Node node, bool can_replace = false);
     Node & getNode(const std::string & name);
+
+    ActionsDAGPtr cloneEmpty() const
+    {
+        auto actions = std::make_shared<ActionsDAG>();
+        actions->max_temporary_columns = max_temporary_columns;
+        actions->max_temporary_non_const_columns = max_temporary_non_const_columns;
+        actions->min_count_to_compile_expression = min_count_to_compile_expression;
+        actions->compile_expressions = compile_expressions;
+
+#if USE_EMBEDDED_COMPILER
+        actions->compilation_cache = compilation_cache;
+#endif
+        return actions;
+    }
+
+    ExpressionActionsPtr linearizeActions() const;
+    void removeUnusedActions(const std::vector<Node *> & required_nodes);
+    void removeUnusedActions();
+    void addAliases(const NamesWithAliases & aliases, std::vector<Node *> & result_nodes);
+
+    void compileFunctions();
 };
 
-using ActionsDAGPtr = std::shared_ptr<ActionsDAG>;
 
 /** Contains a sequence of actions on the block.
   */
```

Comment:
How ExpressionActions related to ActionsDAG? Seems like comment slightly outdated.